src/ol/layer/group.js

```diff
@@ -221,6 +221,10 @@ ol.layer.Group.prototype.getLayerStatesArray = function(opt_states) {
         layerState.extent = ownLayerState.extent;
       }
     }
+    if (layerState.layer.getZIndex() === undefined) {
+      layerState.zIndex = ownLayerState.zIndex;
+      layerState.layer.setZIndex(ownLayerState.zIndex);
```

Comment:
The problem I see is that if the parent group zIndex change, the child zIndex will no longer be computed again, so yes I think it is necessary to change the above code.
I will be in vacations next week, so I will not be able to continue the review. But there are others in the community to continue the review. Maybe @fredj, or @ahocevar will be interested to help having your fix merged.